staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go

```diff
@@ -412,6 +433,94 @@ func (td *typeDiscoverer) discover(t *types.Type, fldPath *field.Path) (*typeNod
 		}
 	}
 
+	// Handle type definitions whose output depends on the rest of type
+	// discovery being complete. In particular, aliases to lists and maps need
+	// iteration, but we don't want to iterate them if the key or value types
+	// don't actually have validations. We also want to handle non-included
+	// types and emit "this page intentionally left blank" comments, but only
+	// if we are emitting other code, lest we end up generating a bunch of
+	// functions whose only purpose is to carry that comment. Lastly, we want
+	// to handle recursive types, but we need to finish discovering the type
+	// before we know if therea re other validations, again so we don't emit
+	// empty functions.
+	if t.Kind == types.Alias {
+		underlying := thisNode.underlying
+
+		switch t.Underlying.Kind {
+		case types.Slice, types.Array:
+			// Validate each value.
+			if elemNode := underlying.node.elem.node; elemNode == nil {
+				if hasValidations(underlying.node) {
+					thisNode.typeValidations.AddComment(fmt.Sprintf(
+						"NOTE: Not iterating this list's values: %v is in a non-included package.", underlying.childType.Elem.Name.Name))
+				}
+			} else {
+				// If the value type is a named type, call the validation
+				// function for each element.
+				if funcName := elemNode.funcName; funcName.Name != "" {
+					// We only need the iteration function if the underlying
+					// type has validations, otherwise it is noise.
+					if hasValidations(underlying.node) {
+						v, err := validators.ForEachVal(fldPath, underlying.childType,
+							validators.Function("iterateListValues", validators.DefaultFlags, funcName))
```

Comment:
This approach is so great.

Very minor thing I'll go ahead and mention: Looks like "iterateListValues" is a placeholder tag name since we don't need one here?  Maybe comment on this? When reading the call site it took me a moment to grok what this string represented.

Reply:
yeah, it isn't actually used anywhere, but it was useful when debugging :)